examples/test-player.rs

```diff
@@ -0,0 +1,465 @@
+        self.shutdown_receiver.recv().unwrap();
+    }
+    pub fn play(&self) {
+        self.sender.send(PlayerEvent::StartPlayback).unwrap();
+    }
+}
+
+struct ImageGenerator {
+    current_frame_receiver: mpsc::Receiver<PlanarYCbCrImage>,
+    current_image: Option<PlanarYCbCrImage>,
+}
+
+impl webrender::ExternalImageHandler for ImageGenerator {
+    fn lock(&mut self, _key: ExternalImageId, channel_index: u8) -> webrender::ExternalImage {
+        if let Ok(v) = self.current_frame_receiver.try_recv() {
+            self.current_image = Some(v);
+        };
+        if let None = self.current_image {
+            panic!("Don't have a current image!");
+        }
+        webrender::ExternalImage {
+            u0: 0.0,
+            v0: 0.0,
+            u1: 1.0,
+            v1: 1.0,
+            source: webrender::ExternalImageSource::RawData(
+                self.current_image
+                    .as_ref()
+                    .map(|p| p.pixel_data(channel_index))
+                    .unwrap(),
+            ),
+        }
+    }
+    fn unlock(&mut self, _key: ExternalImageId, _channel_index: u8) {}
+}
+
+const EXTERNAL_VIDEO_IMAGE_ID: ExternalImageId = ExternalImageId(1);
+
+struct App {
+    image_handler: Option<Box<webrender::ExternalImageHandler>>,
+    frame_receiver: mpsc::Receiver<Vec<PlanarYCbCrImage>>,
+    current_frame_sender: mpsc::Sender<PlanarYCbCrImage>,
+    frame_queue: Vec<PlanarYCbCrImage>,
+    y_channel_key: Option<ImageKey>,
+    cb_channel_key: Option<ImageKey>,
+    cr_channel_key: Option<ImageKey>,
+    last_frame_id: u32,
+    player_wrapper: PlayerWrapper,
+}
+
+impl App {
+    fn new(bytes: Vec<u8>, mime: &'static str) -> App {
+        // Channel for frames to pass between Gecko and player.
+        let (frame_sender, frame_receiver) = mpsc::channel();
+        // Channel for the current frame to be sent between the main
+        // render loop and the external image handler.
+        let (current_frame_sender, current_frame_receiver) = mpsc::channel();
+        let handler = Box::new(ImageGenerator {
+            current_frame_receiver: current_frame_receiver,
+            current_image: None,
+        });
+
+        let app = App {
+            image_handler: Some(handler),
+            frame_receiver: frame_receiver,
+            frame_queue: vec![],
+            current_frame_sender: current_frame_sender,
+            y_channel_key: None,
+            cb_channel_key: None,
+            cr_channel_key: None,
+            last_frame_id: 0,
+            player_wrapper: PlayerWrapper::new(bytes, mime, frame_sender),
+        };
+
+        app
+    }
+    fn shutdown(&self) {
+        self.player_wrapper.shutdown();
+    }
+    fn play(&self) {
+        self.player_wrapper.play();
+    }
+}
+
+fn create_or_update_planar_image(
+    api: &RenderApi,
+    resources: &mut ResourceUpdates,
+    image_key: &mut Option<ImageKey>,
+    plane: Plane,
+    channel_index: u8,
+) {
+    match image_key {
```

Comment:
nit: indentation

Reply:
hmm, `cargo fmt` put that there. Note it's a free function, not a function on a struct. I'll move it into App, since it doesn't really belong in a global scope. Then it would end up indented anyway.